File: app/src/main/java/com/mohsenoid/rickandmorty/domain/Repository.java
package com.mohsenoid.rickandmorty.domain;

import com.mohsenoid.rickandmorty.data.DataCallback;
import com.mohsenoid.rickandmorty.domain.entity.CharacterEntity;
import com.mohsenoid.rickandmorty.domain.entity.EpisodeEntity;

import java.util.List;

public interface Repository {

    void queryEpisodes(int page, DataCallback<List<EpisodeEntity>> callback);

    void queryCharactersByIds(List<Integer> characterIds, DataCallback<List<CharacterEntity>> callback);

    void queryCharacterDetails(int characterId, DataCallback<CharacterEntity> callback);

    void killCharacter(int characterId, DataCallback<CharacterEntity> listDataCallback);
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/domain/Repository.kt
package com.mohsenoid.rickandmorty.domain

import com.mohsenoid.rickandmorty.domain.model.ModelCharacter
import com.mohsenoid.rickandmorty.domain.model.ModelEpisode
import com.mohsenoid.rickandmorty.domain.model.PageQueryResult
import com.mohsenoid.rickandmorty.domain.model.QueryResult

interface Repository {

    suspend fun getEpisodes(page: Int): PageQueryResult<List<ModelEpisode>>

    suspend fun getCharactersByIds(characterIds: List<Int>): QueryResult<List<ModelCharacter>>

    suspend fun getCharacterDetails(characterId: Int): QueryResult<ModelCharacter>

    suspend fun killCharacter(characterId: Int): QueryResult<ModelCharacter>
}
